### PR TITLE
Move Pipelined to Flat Rate Enforcement

### PR DESCRIPTION
**Pull request by koolzz**
# DO NOT MERGE, review and once approved I'll cherry pick to https://github.com/magma/magma/pull/5424

Signed-off-by: Nick Yurchenko <koolzz@fb.com>

## Summary

Convert to flat rate based enforcement
 - Remove all delta calculation maps from enf stats
 - Simplify rule removal to remove rules once we successfully reported flows to sessiond

## Test Plan
 - Unit tests pass 
 - CWF integ passes
 - TerraVM sort of passes (other issues present)
## Additional Information

- [ ] This change is backwards-breaking

Some quick metrics before this pr
```
(python) vagrant@magma-dev:~/magma/lte/gateway/python/scripts$ ./pipelined_cli.py enforcement stress_test_grpc  --test_iterations=2 --attaches_per_sec=100
/home/vagrant/build/python/lib/python3.5/site-packages/scapy/config.py:411: CryptographyDeprecationWarning: Python 3.5 support will be dropped in the next release of cryptography. Please upgrade your Python.
  import cryptography
WARNING: DO NOT USE ON PRODUCTION SETUPS
Attach every ~0.01 seconds
QOS Enabled
Starting iteration 0 of attach/detach requests
Starting attaches
Finished 600 attaches in 11.089374 seconds
Actual attach rate = 54 UEs per sec
Starting detaches
Finished 600 detaches in 6.548913 seconds
Actual detach rate = {0} UEs per sec 92
Starting iteration 1 of attach/detach requests
Starting attaches
Finished 600 attaches in 10.592679 seconds
Actual attach rate = 57 UEs per sec
Starting detaches
Finished 600 detaches in 6.518788 seconds
Actual detach rate = {0} UEs per sec 92
(python) vagrant@magma-dev:~/magma/lte/gateway/python/scripts$ ./pipelined_cli.py enforcement stress_test_grpc  --test_iterations=2 --attaches_per_sec=100
```

After ->
```
/home/vagrant/build/python/lib/python3.5/site-packages/scapy/config.py:411: CryptographyDeprecationWarning: Python 3.5 support will be dropped in the next release of cryptography. Please upgrade your Python.
  import cryptography
WARNING: DO NOT USE ON PRODUCTION SETUPS
Attach every ~0.01 seconds
QOS Enabled
Starting iteration 0 of attach/detach requests
Starting attaches
Finished 600 attaches in 8.591473 seconds
Actual attach rate = 70 UEs per sec
Starting detaches
Finished 600 detaches in 6.66619 seconds
Actual detach rate = {0} UEs per sec 90
Starting iteration 1 of attach/detach requests
Starting attaches
Finished 600 attaches in 10.006942 seconds
Actual attach rate = 60 UEs per sec
Starting detaches
Finished 600 detaches in 6.828944 seconds
Actual detach rate = {0} UEs per sec 88
```

### CODE REVIEW

File: lte/gateway/python/magma/pipelined/app/enforcement_stats.py [hunk 1]
```diff
@@ -496,81 +485,60 @@ def _update_usage_from_flow_stat(self, current_usage, flow_stat):
...
-                    stat_sid, rule_version, e)
+                    'Failed to delete rule %s for subscriber %s ('
+                    'version: %s): %s', record.rule_id,
+                    record.sid, record.rule_version, e)
 
-        self.last_usage_for_delta = self._delta_usage_maps(self.total_usage,
-            deleted_flow_usage)
 
-    def _old_flow_stats(self, stats_msgs):
+    def _old_records(self, records):
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/enforcement_stats.py:510]: can you merge _old_records() and _delete_old_flows(). I dont see much value separating these two functons.

File: lte/gateway/python/magma/pipelined/app/enforcement_stats.py [hunk 2]
```diff
@@ -80,8 +82,6 @@ def __init__(self, *args, **kwargs):
         self._clean_restart = kwargs['config']['clean_restart']
         self._redis_enabled = kwargs['config'].get('redis_enabled', False)
         # Store last usage excluding deleted flows for calculating deltas
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/enforcement_stats.py:84]: can you also remove the comment.

File: lte/gateway/python/magma/pipelined/app/enforcement_stats.py [hunk 3]
```diff
@@ -365,30 +363,12 @@ def _handle_flow_stats(self, stats_msgs):
...
-        self.total_usage = current_usage
-
-        # Append any records which we couldn't send to session manager earlier
-        delta_usage = _merge_usage_maps(delta_usage, self.failed_usage)
-        self.failed_usage = {}
-
         # Send report even if usage is empty. Sessiond uses empty reports to
         # recognize when flows have ended
-        self._report_usage(delta_usage)
+        self._report_usage(current_usage)
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/enforcement_stats.py:368]: _report_usage() definition still uses  delta_usage as parameter. lets rename it.

File: lte/gateway/python/magma/pipelined/app/enforcement_stats.py [hunk 4]
```diff
@@ -459,17 +443,22 @@ def _update_usage_from_flow_stat(self, current_usage, flow_stat):
...
         rule_version = _get_version(flow_stat)
         if not rule_version:
             rule_version = 0
+
+        key += "|" + str(rule_version)
+
+        record = current_usage[key]
+        record.rule_id = rule_id
+        record.sid = sid
+
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/enforcement_stats.py:461]: do we still need map here? list of stats record enough, no?

Comment (koolzz) [reply lte/gateway/python/magma/pipelined/app/enforcement_stats.py:461]: multiple flows can match to one record (usually at least 2 flows in the policyrule flowlist) so we still need map

**Review by pshelar — APPROVED** [review]
LGTM

**Pull request closed without merging**